feat(computability): prove equivalence of NFAs and regular expressions

RussellEmerine opened the pull request:
Provide a way to convert NFAs to regular expressions and vice versa, and prove that the conversions accept the same language.

TODO: statement of some theorems, style, adding @[simp] annotations appropriately, conversions between DFAs and regular expressions

RussellEmerine committed:
feat(library): prove NFAs are equivalent to regular expressions

This adds a working proof that NFAs are equivalent to regular
expressions.  This is *not* meant to be committed into the actual
library yet, rather just so people can see my work so far. I will soon
fix style things and add comments on parts that I'm unsure about.

RussellEmerine committed:
doc(computability): documentation, TODOs, and questions

Add documentation, TODOs, and questions about my proof. Still a WIP.

RussellEmerine committed:
style(computability): fix style and remove intermediate simp

Use standard brace placement in tactic mode. Replace intermediate simp
with rw and simp only.

RussellEmerine committed:
style(computability): fix long lines and braces

Fix all style issues

eric-wieser commented on src/computability/GNFA.lean:
```diff
@@ -0,0 +1,675 @@
+Given an equivalence between `σ` and `τ`, convert an NFA with state type `σ` into the corresponding
+NFA with state type `τ`.
+
+TODO: possibly move to computability/NFA
```
Since this is your first PR and it's rather large; I recommend you split this definition and the lemma below into a second PR (adding it to the NFA file), and then come back to this one once that's merged.

eric-wieser commented on src/computability/regular_expression_to_NFA/star.lean:
```diff
@@ -0,0 +1,271 @@
+ * probably move this into computability/regular_expression
+ * maybe change the order to `pow n * r`?
+-/
+def pow (r : regular_expression α) : ℕ → regular_expression α
```
Is `monoid (regular_expression a)` true?

eric-wieser replied:
Ah no, because regexes have no useful meaning of equality.

eric-wieser replied:
I've moved this definition to #14261 and added a lemma.

eric-wieser commented on src/computability/regular_expression_to_NFA/star.lean:
```diff
@@ -0,0 +1,271 @@
+end
+
+theorem star_iff_pow {r : regular_expression α} {x} :
+  x ∈ r.star.matches ↔ (∃ (n : ℕ), x ∈ (r ^ n).matches) :=
```
I think this follows trivially from `language.star_eq_supr_pow`

RussellEmerine replied:
Thank you for all your feedback, Eric. I will be making changes and keeping an eye on #14261 .

eric-wieser replied:
It's now merged :)

RussellEmerine committed:
feat(computability/regular_expressions): regex power operator

Merge updates to master, including a regular expression power operator
that will be used in the proof of correctness for the star case of
regular expression to NFA

RussellEmerine committed:
substituted a simplification

RussellEmerine committed:
Merge branch 'master' into RussellEmerine/DFA_equivalent_regular_expression

RussellEmerine committed:
feat(computability/GNFA)

Refactored GNFA proofs to use fintype.induction_empty_option instead of fin.

RussellEmerine committed:
Updated imports
